backend/de.metas.business/src/main/java/org/adempiere/mm/attributes/api/impl/AttributesBL.java

```diff
@@ -212,13 +214,20 @@ public boolean hasAttributeAssigned(@NonNull final ProductId productId, @NonNull
 	}
 
 	@Override
-	public boolean isMandatoryOnReceipt(@NonNull final ProductId productId, @NonNull final AttributeId attributeId)
+	public boolean isMandatoryOn(@NonNull final ProductId productId,
+			@NonNull final AttributeId attributeId,
+			@NonNull AttributeSourceDocument attributeSourceDocument)
 	{
 		final AttributeSetId attributeSetId = productsService.getAttributeSetId(productId);
+		final Function<AttributeSetAttribute, OptionalBoolean> mandatoryOnFunction = attributeSourceDocument.isMaterialReceipt() ?
+				AttributeSetAttribute::getMandatoryOnReceipt :
+				AttributeSetAttribute::getMandatoryOnManufacturing;
+
 		final Boolean mandatoryOnReceipt = attributesRepo.getAttributeSetAttributeId(attributeSetId, attributeId)
```

Comment:
this is pretty weird. to make it a bit clearer:
* inline the mandatoryOnFunction
*rename mandatoryOnRecpit to simply `mandatory`

.... but the best would be to have something like:
```java
final AttributeSetAttribute attribute = attributesRepo.getAttributeSetAttributeId(attributeSetId, attributed);
final Boolean mandatory;
if (attributeSourceDocument.isReciept()) {
mandatory = attribute.getMandatoryOnReciept().orElse(null);
}
else if (attributeSourceDocument.isManufacturing()) {
mandatory = attribute.getMandatoryOnManufacturing().orElse(null);
}
else if (attributeSourceDocument.isPicking()) {
mandatory = attribute.getMandatoryOnPicking().orElse(null);
}
else {
throw new AdempiereException("Unknown: "+attributeSourceDocument);
}
```